### PR TITLE
Support for JWT Bearer 

### PR DESCRIPTION
**Pull request by kralos**
https://tools.ietf.org/html/draft-ietf-oauth-jwt-bearer-12

Please note, http://tools.ietf.org/html/draft-ietf-oauth-v2-20#section-4.5 refers to a URI which is defined as a URL or URN http://www.rfc-editor.org/rfc/rfc3305.txt

### CODE REVIEW

**Comment by kralos** [comment]
See Issue #78

**Comment by Spomky** [comment]
Looks good.
Ping @alanbem

**Comment by Turneliusz** [comment]
@Spomky @alanbem any progress on that? Would love to use

**Comment by michaelfeinbier** [comment]
@alanbem It would be great to merge this!

**Comment by Spomky** [comment]
@Ener-Getick can you merge and tag this PR?

File: lib/OAuth2.php [hunk 1]
```diff
@@ -1001,8 +997,7 @@ protected function grantAccessTokenExtension(IOAuth2Client $client, array $input
         if (!($this->storage instanceof IOAuth2GrantExtension)) {
             throw new OAuth2ServerException(self::HTTP_BAD_REQUEST, self::ERROR_UNSUPPORTED_GRANT_TYPE);
         }
-        $uri = filter_var($inputData["grant_type"], FILTER_VALIDATE_URL);
```
Comment (Ener-Getick) [inline lib/OAuth2.php]: why did you remove this ?

Comment (Turneliusz) [inline lib/OAuth2.php]: URN is valid URL so should work with it

Comment (Spomky) [inline lib/OAuth2.php]: You may check both URI and URN.

A regex pattern for URN could be something like `^urn:[a-z0-9][a-z0-9-]{0,31}:[a-z0-9()+,\-.:=@;$_!*'%/?#]+$`

File: lib/OAuth2.php [hunk 2]
```diff
@@ -823,12 +823,8 @@ public function grantAccessToken(Request $request = null)
                 $stored = $this->grantAccessTokenRefreshToken($client, $input);
                 break;
             default:
-                if (filter_var($input["grant_type"], FILTER_VALIDATE_URL)) {
```
Comment (Ener-Getick) [inline lib/OAuth2.php]: Maybe it would be better to change this check to keep the uri verification:
```php
if (filter_var($input["grant_type"], FILTER_VALIDATE_URL) || 0 === strpos($input["grant_type"], 'urn')) {
```

File: lib/OAuth2.php (hunk 1 above)
Comment (Ener-Getick) [inline lib/OAuth2.php]: Wouldn't it be better to do:
```php
if (false === strpos($inputData['grant_type'], 'urn') {
    $uri = filter_var($inputData["grant_type"], FILTER_VALIDATE_URL);
}
```
?

**Comment by Ener-Getick** [comment]
Can you add a test case to be sure this works as expected please?

File: lib/OAuth2.php (hunk 1 above)
Comment (Ener-Getick) [inline lib/OAuth2.php]: @Turneliusz actually, it doesn't work https://secure.php.net/manual/fr/filter.filters.validate.php#110411

Comment (Turneliusz) [inline lib/OAuth2.php]: @Ener-Getick good to know, thx

**Comment by kralos** [comment]
Added test case and restored the `FILTER_VALIDATE_URL` checks etc for non-urn grant extensions since PHP thinks they are invalid

**Comment by Turneliusz** [comment]
Looking forward, thanks @kralos !

**Comment by Ener-Getick** [comment]
Merged in https://github.com/FriendsOfSymfony/oauth2-php/commit/fa2aecb1fca2a03fd5f9aca19fe9adb9dfff928c, thank you @kralos !

I'll wait a bit to be completely sure this works as expected. Can you test it in your projects please?
Don't hesitate to ping me if I forget it for more than a week ;-)
